model.py

```diff
@@ -0,0 +1,209 @@
+        decay=step.hyperparams['lr_decay'],
+        momentum=step.hyperparams['momentum']
+    )
+
+
+def create_feed_dict(step: TensorflowV1ModelStep, data_inputs, expected_outputs):
+    expected_outputs_length = []
+    for expected_output in expected_outputs:
+        expected_outputs_length.append(len(expected_output))
+
+    # shape: (batch_size)
+    expected_outputs_length = np.array(expected_outputs_length, dtype=np.int32)
+
+    return {
+        step['expected_outputs_length']: expected_outputs_length
+    }
+
+
+def to_numpy_metric_wrapper(metric_fun):
+    def metric(data_inputs, expected_outputs):
+        return metric_fun(np.array(data_inputs)[..., 0], np.array(expected_outputs)[..., 0])
+
+    return metric
+
+
+class SignalPredictionPipeline(Pipeline):
+    BATCH_SIZE = 5
+    LAMBDA_LOSS_AMOUNT = 0.003
+    OUTPUT_DIM = 2
+    INPUT_DIM = 2
+    HIDDEN_DIM = 12
+    LAYERS_STACKED_COUNT = 2
+    LEARNING_RATE = 0.006
+    LR_DECAY = 0.92
+    MOMENTUM = 0.5
+    OUTPUT_SIZE = 5
+    EPOCHS = 150
+
+    def __init__(self):
+        super().__init__([
+            TensorflowV1ModelStep(
+                create_graph=create_graph,
+                create_loss=create_loss,
+                create_optimizer=create_optimizer,
+                create_feed_dict=create_feed_dict
+            ).set_hyperparams(HyperparameterSamples({
+                'batch_size': self.BATCH_SIZE,
+                'lambda_loss_amount': self.LAMBDA_LOSS_AMOUNT,
+                'output_dim': self.OUTPUT_DIM,
+                'output_size': self.OUTPUT_SIZE,
+                'input_dim': self.INPUT_DIM,
+                'hidden_dim': self.HIDDEN_DIM,
+                'layers_stacked_count': self.LAYERS_STACKED_COUNT,
+                'learning_rate': self.LEARNING_RATE,
+                'lr_decay': self.LR_DECAY,
+                'momentum': self.MOMENTUM
+            })),
+        ])
+
+
+if __name__ == '__main__':
+    exercise = 4
+    # We choose which data function to use below, in function of the exericse.
+    if exercise == 1:
+        generate_x_y_data = generate_x_y_data_v1
+    if exercise == 2:
+        generate_x_y_data = generate_x_y_data_v2
+    if exercise == 3:
+        generate_x_y_data = generate_x_y_data_v3
+    if exercise == 4:
+        generate_x_y_data = generate_x_y_data_v4
+
+    pipeline = DeepLearningPipeline(
+        SignalPredictionPipeline(),
+        validation_size=0.15,
+        batch_size=SignalPredictionPipeline.BATCH_SIZE,
+        batch_metrics={'mse': to_numpy_metric_wrapper(mean_squared_error)},
+        shuffle_in_each_epoch_at_train=True,
+        n_epochs=SignalPredictionPipeline.EPOCHS,
+        epochs_metrics={'mse': to_numpy_metric_wrapper(mean_squared_error)},
+        scoring_function=to_numpy_metric_wrapper(mean_squared_error)
+    )
+
+    data_inputs, expected_outputs = generate_x_y_data(isTrain=True, batch_size=SignalPredictionPipeline.BATCH_SIZE)
```

Comment:
I'm not sure what happens here so the following may be wrong but I think I spotted a problem: 

I think your problem is that I was calling the `generate_x_y_data` function at each epoch. You'd probably need here a train and a test data generator class that lazy loads each batches for the DL Minibatching Pipeline. 

You'd also need to split the data on the time axis like this here if using cross validation: 
![](https://miro.medium.com/max/732/1*obKmc_bTKbUFgcgryhaAnA.png)

So the users could optimize on the validation data with a manual search and manual parameter tuning (no random search), and then call a function to test on the test set or something. So we need to probably change the `generate_x_y_data` method.

Reply:
the cross validation is done in the deep learning pipeline.

Reply:
oh wait you don't have all my changes I actually have a loop that calls generate_x_y_date a few times :/

Reply:
let's change generate_x_y_data_v4 it was confusing

Reply:
also the normalization, and lazy data loading should be done in neuraxle